tools/checked-c-convert/ProgramInfo.cpp

```diff
@@ -634,72 +757,160 @@ ProgramInfo::getVariableHelper( Expr                            *E,
-                fwdDecl = RD;
-                break;
-              }
-            }
-
-            if (fwdDecl) {
-              // We found one! Let's figure out the index that D has in Parent,
-              // then get that decl from fwdDecl and look it up in Variables
-              // by PSL, then return it. 
-              int idx = -1;
-              
-              for (unsigned i = 0; i < Parent->getNumParams(); i++) {
-                const ParmVarDecl *tmp = Parent->getParamDecl(i);
-
-                if (tmp == D) {
-                  idx = i;
-                  break;
-                }
-              }
+    // not. If we are not, then we should find a declaration
+    ParmVarDecl *PD = nullptr;
+    FunctionDecl *funcDefinition = nullptr;
+    FunctionDecl *funcDeclaration = nullptr;
+    // get the function declaration and definition
+    if (D != nullptr && dyn_cast<FunctionDecl>(D)) {
+      funcDeclaration = getDeclaration(dyn_cast<FunctionDecl>(D));
+      funcDefinition = getDefinition(dyn_cast<FunctionDecl>(D));
+    }
+    int parameterIndex = -1;
+    if (PD = dyn_cast<ParmVarDecl>(D)) {
+      // okay, we got a request for a parameter
+      DeclContext *DC = PD->getParentFunctionOrMethod();
+      assert(DC != nullptr);
+      FunctionDecl *FD = dyn_cast<FunctionDecl>(DC);
+      // get the parameter index with in the function.
+      for (unsigned i = 0; i < FD->getNumParams(); i++) {
+        const ParmVarDecl *tmp = FD->getParamDecl(i);
+        if (tmp == D) {
+          parameterIndex = i;
+          break;
+        }
+      }
 
-              assert(idx >= 0);
+      // get declaration and definition
+      funcDeclaration = getDeclaration(FD);
+      funcDefinition = getDefinition(FD);
 
-              const ParmVarDecl *otherDecl = fwdDecl->getParamDecl(idx);
-              I = Variables.find(PersistentSourceLoc::mkPSL(otherDecl, *C));
-              assert(I != Variables.end());
+      assert(parameterIndex >= 0 && "Got request for invalid parameter");
+    }
+    if (funcDeclaration || funcDefinition || parameterIndex != -1) {
+      // if we are asking for the constraint variable of a function
+      // and that function is an external function.
+      // then use declaration.
+      if (dyn_cast<FunctionDecl>(D) && funcDefinition == nullptr) {
+        funcDefinition = funcDeclaration;
+      }
+      // this means either we got a
+      // request for function return value or parameter
+      if (inFunctionContext) {
+        assert(funcDefinition != nullptr && "Requesting for in-context constraints, "
+                                            "but there is no definition for this function");
+        // return the constraint variable
+        // that belongs to the function definition.
+        return getVariable(D, C, funcDefinition, parameterIndex);
+      } else {
+        if (funcDeclaration == nullptr) {
+          // we need constraint variable
+          // with in the function declaration,
```

Comment:
Type: with in-> within.   Please take advantage of the 80-character line limit for comments :).

Reply:
Fixed.